sematic/storage.py

```diff
@@ -1,124 +1,193 @@
-        # Standardizing "Not found" errors across storage backends
-        if "404" in str(e):
-            raise KeyError("{}: {}".format(key, str(e)))
 
-        raise e
-    return file_obj.getvalue()
+    def set(self, key: str, value: bytes):
+        dir_path = os.path.split(key)[0]
+        os.makedirs(os.path.join(get_config().data_dir, dir_path), exist_ok=True)
 
+        with open(os.path.join(get_config().data_dir, key), "wb") as file:
+            file.write(value)
 
-@retry(tries=3, delay=5)
-def get_line_stream(key: str, encoding="utf8") -> Iterable[str]:
-    """Get value from S3 into a stream of text lines.
+    def get(self, key: str) -> bytes:
+        try:
+            with open(os.path.join(get_config().data_dir, key), "rb") as file:
+                return file.read()
+        except FileNotFoundError:
+            raise NoSuchStorageKey(self, key)
 
-    The encoding of the
 
-    See TODO in `set`.
+class S3Storage(Storage):
+    """
+    Implementation of the `Storage` interface for AWS S3 storage. The bucket
+    where to store values is determined by the `AWS_S3_BUCKET` user settings.
     """
-    s3_client = boto3.client("s3")
-
-    try:
-        obj = s3_client.get_object(Bucket=_get_bucket(), Key=key)
-        return _bytes_buffer_to_text(obj["Body"].iter_lines(), encoding)
-    except botocore.exceptions.ClientError as e:
-        # Standardizing "Not found" errors across storage backends
-        if "404" in str(e):
-            raise KeyError("{}: {}".format(key, str(e)))
 
-        raise e
+    @memoized_property
+    def _bucket(self) -> str:
+        return get_user_settings(SettingsVar.AWS_S3_BUCKET)
+
+    @memoized_property
+    def _s3_client(self):
+        return boto3.client("s3")
+
+    def set(self, key: str, value: bytes):
+        """Store value in S3"""
+        with io.BytesIO(value) as file_obj:
+            self._s3_client.upload_fileobj(file_obj, self._bucket, key)
+
+    @retry(tries=3, delay=5)
+    def get(self, key: str) -> bytes:
+        """Get value from S3.
+
+        See TODO in `set`.
+        """
+        file_obj = io.BytesIO()
+        try:
+            self._s3_client.download_fileobj(self._bucket, key, file_obj)
+        except botocore.exceptions.ClientError as e:
+            # Standardizing "Not found" errors across storage backends
+            if "404" in str(e):
+                raise NoSuchStorageKey(self, key)
+
+            raise e
+        return file_obj.getvalue()
+
+    def set_from_file(self, key: str, value_file_path: str):
+        """Store value in S3 using the contents of a file
+
+        see TODO in 'set'
+        """
+        with open(value_file_path, "rb") as file_obj:
+            self._s3_client.upload_fileobj(file_obj, self._bucket, key)
+
+    @retry(tries=3, delay=5)
+    def get_line_stream(self, key: str, encoding="utf8") -> Iterable[str]:
+        """Get value from S3 into a stream of text lines.
+
+        The encoding of the
```

Comment:
```suggestion
```

I think it was likely to be something about the encoding being assumed utf8, but then I added it as a function parameter which makes it more obvious.

Reply:
Bump.

Reply:
Done.